src/app.jsx

```diff
@@ -71,20 +89,23 @@ export const Application = () => {
             else if (item.event === 'deleted') {
                 const deleted_name = item.path.slice(item.path.lastIndexOf("/") + 1);
                 setFiles(f => f.filter((res) => { return res.name !== deleted_name }));
-            } else if (item.event === 'created') {
-                const created_name = item.path.slice(item.path.lastIndexOf("/") + 1);
-                const date_modified = item.tag.slice(item.tag.indexOf("-") + 1, item.tag.lastIndexOf("."));
-                if (created_name[0] !== ".") {
-                    setFiles(f => [...f, { name: created_name, type: item.type, modified: date_modified }]);
-                }
-            } else if (item.event === 'changed' || item.event === 'attribute-changed') {
-                const changed_name = item.path.slice(item.path.lastIndexOf("/") + 1);
-                const date_modified = item.tag.slice(item.tag.indexOf("-") + 1, item.tag.lastIndexOf("."));
-                setFiles(f => f.map((file) => {
-                    return file.name === changed_name
-                        ? { name: file.name, type: item.type, modified: date_modified }
-                        : file;
-                }));
+            } else if (item.event === 'created' || item.event === 'changed') {
+                refreshFslist();
+                // const created_name = item.path.slice(item.path.lastIndexOf("/") + 1);
+                // const date_modified = item.tag.slice(item.tag.indexOf("-") + 1, item.tag.lastIndexOf("."));
```

Comment:
I don't commit comments but I leave them in the cases where I'm stuck on an issue, so it's easier to go back to previous code without looking through old commits. If it's better to always remove them let me know!

The issue here is that the `cockpit.channel` send multiple messages per event. For example, saving a file sends 'changed' twice, which runs the refresh function twice instantly which breaks the state

Reply:
This depends on how exactly you change a file. One option is to "open, truncate, rewrite, close", which will send one or two events. But most editors do a different approach: they write a *new* file with the new contents (causing several events), and once that is successful, rename the new file to the old name (causing more events). This is generally the preferred and safer (because "atomic"), method, as you never end up with a half-written file.

What do you mean with "breaks the state"? that usually happens if you do something like `setState({ ... oldState ... })`, as that happens asynchronously. That's why in such cases it's better to use the functional form. But that doesn't seem to happen in this PR?

Of course you should still avoid calling setState() a lot of times in a burst, as that causes flicker and is slow. There are various ways to avoid it, but all of that is more like "optimization", not "breakage". Do you still see wrong results here? If so, let's figure that out. Thanks!

Reply:
Ah, if the UI doesn't show this information yet, then this PR sounds like a no-op? It sounds like you have another PR which adds this functionality. Then I recommend to add that commit to that other PR, and test it there. If this is impractical for some reason, then feel free to land this. But note that this is *not* the way that software should be developed, so please treat that as an exception, not a model how to do things. :grin:

Reply:
Adding this change to the other PR sounds like the best option here, thanks!